db/view/view.cc

```diff
@@ -121,6 +121,11 @@ const column_definition* view_info::view_column(const column_definition& base_de
 
 void view_info::set_base_info(db::view::base_info_ptr base_info) {
     _base_info = std::move(base_info);
+    // Base-dependent view info got overridden, so the cached select statement
+    // should be reset to the current schema - it's done by dropping the current
+    // value and calling the function which regenerates it.
+    _select_statement = nullptr;
+    select_statement();
```

Comment:
I don't think there's any reason to call select_statement() here. The information will be recalculated the next time someone calls select_statement(), there is no particular reason (unless I'm missing something...) to do it now.
It is the "_select_statement = nullptr" line which is the important one. Right? Or am I missing something?

Reply:
There is a reason - all view updates are generated with a particular schema version. Even if schema is altered in the middle of the operations, all updates are generated with the same schema snapshot to remain consistent. That's why I synchronously update the statement to, to keep it atomic. Otherwise we'll end up with mismatched schema if it gets altered in the middle of view update generation

Reply:
@psarna I don't understand how this is a response to my comment -
What I meant was that as far as I can tell, nobody uses `_select_statement` directly - callers call `select_statement()`, so if `_select_statement` is set to nullptr as you do, then the next call to `select_statement()` will recalculate it. What I'm asking is whether there is any point of recalculating right now and not doing anything with the result.

Reply:
The reason is that right now recalculation always takes the newest schema from memory - peek at the code, it uses base id to fetch the newest schema. The newest schema in memory does not necessarily match the schema we keep in base-dependent view info, and we need this assumption to hold, due to the fact that existing code already assumes it.

Reply:
I don't understand :-(
If calling `select_statement()` at the very last line of `set_base_info()`, won't it work when called _after_ `set_base_info()` concludes? I don't understand what I am missing... 
Did you actually see this line was necessary in some test, and the`_select_statement = nullptr` wasn't enough?

Reply:
It would work right after set_base_info concludes, but before any preemption point. It works like that:
1. We create a view_updates instance with a particular base schema; this instance uses the same schema to generate updates; the whole generation process has many preemption points
2. We already assume that schema from this select_statement thing matches the schema used for view updates.

If I don't regenerate the statement right away, we risk the following:
1. We start generating view updates with given base schema (newest schema at this time point)
2. Preemption happens, we update the schema because it was altered
3. We call select_statement() and end up with this statement having a schema that doesn't match schema from (1.), breaking our assumption

Reply:
So, regarding your 1,2,3 scenario - what happens if at point 2 you call your set_base_info() function and it recalculates select_statement() immediately - doesn't this cause exactly the same problem at point 3?

Reply:
Hm, I don't know, but perhaps you're right and the race is still there. This schema and view info management operates on shared pointers in order to have snapshots of given schema, so **maybe** it's somehow guaranteed that a new schema won't call set_base_info on an old view info, but rather on some new object, but I'm speculating here - I have no idea :(

Reply:
I'm sorry, I can't help you figure out the answer because I still don't understand the problem ;-)
I understood the "_select_statement = nullptr" command - we had some object (_select_statement) which referred to some old data and we need to drop this object because it's no longer usable. What I don't understand is why you are saying that this is a "race" and it matters when we calculate the new value.
Do you have any sort of reproducer for this issue?

All of this being said, I don't think there's anything **wrong** with the `select_statement()` call - I don't think it does anything wrong - we need to calculate the new value sometime so we might as well just do it immediately. I'm just worried that the fact you thought it's necessary and I thought it wasn't means one of us - or both - don't understand what this patch really does - and that's a reason to worry ;-)

Reply:
My reasoning only makes sense under the assumption that nobody is going to call `set_base_info` on a view_info instance when view updates are generated. I have no evidence that it holds, it likely doesn't, so this discussion is now purely theoretical :) You already showed that if my assumption doesn't hold, select_statement can still hold 
a schema that doesn't match the one we use for view updates: https://github.com/scylladb/scylla/pull/10040#discussion_r800544569 .
So let's start over, assuming that nobody is going to call `set_base_info` on a view_info that undergoes view update generation. What I want to ensure now is that the base schema held in the `_select_statement` attribute is exactly the same as the base schema used for view update generation, because otherwise we end up with bug #10026.
My observation is that each time a user alters the schema of a table, the newest version of this schema in memory calls `set_base_info` for each of its views, updating their `_base_info` attributes. Now, if we immediately regenerate the `_select_statement`, it's guaranteed to be generated from the same newest schema for that table, because there wasn't any preemption point before the regeneration, and thus we use the same newest schema that we used for `set_base_info`. However, if we just reset the `_select_statement` to nullptr and lazily wait for it to be regenerated, it's no longer true that we'll always have matching schemas in `_base_info` and `_select_statement`. That's because we can have a preemption point after we called `_set_base_info` to schema S1 and reset the  `_select_statement`, and meanwhile the schema could already change to S2. Later, when we want to use `_select_statement` during view update generation, we'll regenerate it lazily from the newest schema S2, which doesn't match schema S1, from which view updates are generated.

Reply:
Ok, I think I finally see your point. `table::set_schema()` does:
```c++
    _schema = std::move(s);

    for (auto&& v : _views) {
        v->view_info()->set_base_info(
            v->view_info()->make_base_dependent_view_info(*_schema));
    }
```

And you want to use this `_schema` also to calculate the `_select_statement` - not to wait until later when `_schema` is potentially different.

I'm not 100% sure I understood everything and this has all the bases covered, but I'm pretty sure it's better than what we had before, so let's commit this patch.

Reply:
@nyh actually, @eliransin is taking it for a test spin (if I understood the discussion from #10026 correctly), so we may just as well wait and see if this patch fixes at least the particular test that failed. If it doesn't, then there's probably no point in queuing it and we should continue investigating

Reply:
I still don't understand why calculating select_statement() immediately instead of later is necessary, whereas before this patch it was always delayed and things worked. 
But the sad fact is that this immediate setting breaks tests when they try to delete a table which has an index. I don't know why, but I give up. I'm guessing this code still has races but it has simply become impossible to understand, and we should start by committing something that fixes tests, not fixes 1 test and breaks 20 others. CC @eliransin